matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m

```diff
@@ -0,0 +1,172 @@
+function [pose,map,y,muFeat,particleSet,t,meas_ind,newfeature,centroid_particles]=fastSLAM_func(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize)
+totalFeatures=MapSize;R=MeasurementNoise;Q=MotionNoise;
+
+%initialize the Fast Slam parameters
+[timeInterval,finalTime,noFeatures,map,pose,totalParticles,muFeat,muFeatPred,covFeat,covFeatPred,newfeature,w,u,timeStamp,rmax,thmax]...
+            =initializeFastSLAM(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize);
+%% Fast SLAM LOOP
+for t=2:finalTime/timeInterval
+    timeStamp(t)=timeStamp(t-1)+timeInterval;
+    pose(:,t)=motionUpdate_fs(pose(:,t-1),u,Q,timeInterval);
+    meas_ind=featureSearch_fs(noFeatures,map,pose(:,t),rmax,thmax); 
+    y=[];
+    for feature_index = meas_ind       
+        y_new=measurementUpdate_fs(map(:,feature_index),pose(:,t),R);
+        y=[y,y_new];     
+    end
+
+    for particle=1:totalParticles
+       particleSetP(:,particle)=motionUpdate_fs(particleSet(:,particle),u,Q,timeInterval);
+       clear yw hmuw Qw;
+       for j=1:length(meas_ind)
+           i = meas_ind(j);
+           [muFeatPred(:,i,particle),covFeatPred(:,:,i,particle),Hmu]=...
+               ekf_fs(particleSetP(:,particle),y(:,j),muFeat(:,i,particle),...
+               covFeat(:,:,i,particle),R,newfeature(i));
+           yw(2*(j-1)+1:2*j) = y(:,j);
+           hmuw(2*(j-1)+1:2*j) = Hmu;
+           Qw(2*(j-1)+1:2*j,2*(j-1)+1:2*j) = 2*covFeatPred(:,:,i,particle);
```

Comment:
Yes. I did not understand that either. Ideally, it should have been..
H_feat*covFeatPred(:,:,i,particle)*H_feat'+measurementNoise;
where H_feat is the measurement jacobian with respect to the predicted features at that time instant and CovFeatPred is the covariance of that particle at that time step of a particular feature. 

I have made the changes in the function and it works. I shall push the updated code once the review is complete. Nice that you noticed..;-)